### PR TITLE
Improve SMTP authentication and Fix user creation bugs 

### LINKED ISSUE #14102: LDAP with custom root CA
- Gitea version (or commit ref):  1.13.0 - Deployed using official helm chart to k3s cluster
- Operating system: Docker image/Alpine v3.12
  
- Database (use `[x]`):
  - [ ] PostgreSQL
  - [x] MariaDB
  - [ ] MSSQL
  - [ ] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [ ] Yes (provide example URL)
  - [x] No
- Log:
```
2020/12/22 00:13:56 ...es/auth/ldap/ldap.go:241:SearchEntry() [E] LDAP Connect error, <LDAP.ADDR>:LDAP Result Code 200 "Network Error": x509: certificate signed by unknown authority
```

## Description
Gitea doesn't provide a method for allowing admins to inject custom root CAs for LDAPS authentication.  Most web applications that I've used which leverage LDAP allow for admins to paste in a root CA for doing ldap authentication.

## Scope
If you can add your CA cert to the root trust store for the OS, then you're golden.  For regular VM/bare metal installs, this is almost a non-issue.  In most cases, those hosts would likely be joined to your domain (IPA or AD) and likely have the root CA cert for your domain already added.

For docker/kubernetes, this is a non-starter.  You have to either:
- Run the commands (below) manually in the container
  - The volume where the updated trust store (`/etc/ssl/certs/ca-certificates.crt`) is ephemeral and subsequently not 'durable'
- Do some witchcraft and `volumeMount` over the existing system cert store
  - This means that the admin has to maintain the volume and rebuild the cert store every time the image gets updated.  Unsustainable.
- Rebuild the official gitea with your certs baked in
  - Again, unsustainable.  Barrier of entry too high just to get LDAPS.
- Disable TLS verification in the authentication configuration (what I suspect most people are doing)
  - Might as well disable TLS entirely

## Workaround
Based off of the alpine linux image, one can update the root store like so:
```
# source: https://unix.stackexchange.com/questions/464484/install-self-signed-certificate-to-alpine-linux
curl http://ipa.server/ipa/config/ca.crt > /usr/local/share/ca-certificates/iparoot.ca
update-ca-certificates
```
But as mentioned above, (for docker/kube at least), this is very fragile.

This issue is not new as it's been alluded to in #6335 and #13396.  

Thoughts?

### PR DESCRIPTION
**Pull request by zeripath**
This PR has two parts:

Improvements for SMTP authentication:

* Default to use SMTPS if port is 465, and allow setting of force SMTPS.
* Always use STARTTLS if available
* Provide CRAM-MD5 mechanism
* Add options for HELO hostname disabling
* Handle application specific password response as a failed user login
instead of as a 500.

Close #16104

Fix creation of new users:

* A bug was introduced when allowing users to change usernames which
prevents the creation of external users.
* The LoginSource refactor also broke this page.

Close #16104

Signed-off-by: Andrew Thornton <art27@cantab.net>

### CODE REVIEW

**Comment by zeripath** [comment]
Agh I need to update the docs and the command too. Will add those tomorrow.

**Comment by lafriks** [comment]
Imho it would be better to use single pem file for both certificate and key, something like this:
https://gist.github.com/ukautz/cd118e298bbd8f0a88fc#file-cert_load-go-L18

**Comment by zeripath** [comment]
I guess the trouble is we haven't been using that kind of thing before. This PR as it currently stands means that everywhere you have certificates and keys in Gitea you know you need to have them separate.

**Comment by lafriks** [comment]
Oh, had not noticed that. Having them in single file makes things way easier imho but I will not block it either way :)

**Comment by zeripath** [comment]
So the one questionable part of this PR is the option to provide the path to the key/cert combo - this is not restricted to the data path and the path can be anything.

This page is only accessible to the administrator, but perhaps it might be better to make it an option to use a cert/key specifically named for the login_source as in a checked bool option instead.

**Comment by lafriks** [comment]
Why not store certificate/key in database in text/blob field? This would also make it possible later to move it to gitea secrets

File: services/auth/source/ldap/source_search.go [hunk 1]
```diff
@@ -103,26 +108,47 @@ func (ls *Source) findUserDN(l *ldap.Conn, name string) (string, bool) {
...
-		InsecureSkipVerify: ls.SkipVerify,
+	tlsConfig := &tls.Config{
+		ServerName:         source.Host,
+		InsecureSkipVerify: source.SkipVerify,
 	}
-	if ls.SecurityProtocol == SecurityProtocolLDAPS {
-		return ldap.DialTLS("tcp", fmt.Sprintf("%s:%d", ls.Host, ls.Port), tlsCfg)
+
+	if source.UseCertificate {
+		certFile := filepath.Join(setting.AppDataPath, "login-sources", source.Name+".cert.pem")
```
Comment (lafriks) [inline services/auth/source/ldap/source_search.go]: It makes hard to use it, especially in dockerized environment

Comment (zeripath) [reply services/auth/source/ldap/source_search.go]: but this is just a subdirectory of the data path?

Comment (lafriks) [reply services/auth/source/ldap/source_search.go]: Yes, I know but you can't really manage it though docker for example, just from file system

**Comment by zeripath** [comment]
> Why not store certificate/key in database in text/blob field? This would also make it possible later to move it to gitea secrets

I guess we could encrypt them with the secret key?

My main worry is that it's going to need another huge field that we will have to get out of the db every time someone tries to login and potentially other times too. So it would need to be another table where it could be cached etc.

---

it should also be noted that this option is only for providing a client certificate - this is not for verifying the host certificate.

File: services/auth/source/ldap/source_search.go (hunk 1 above)
Comment (zeripath) [reply services/auth/source/ldap/source_search.go]: but this is like the rest of our certificates and keys.

**Comment by lafriks** [comment]
> > Why not store certificate/key in database in text/blob field? This would also make it possible later to move it to gitea secrets
> 
> I guess we could encrypt them with the secret key?
> 
> My main worry is that it's going to need another huge field that we will have to get out of the db every time someone tries to login and potentially other times too. So it would need to be another table where it could be cached etc.
> 
> it should also be noted that this option is only for providing a client certificate - this is not for verifying the host certificate.

Or maybe wait for secrets feature to be added

**Comment by zeripath** [comment]
So you suggest removing this part of the PR?

**Comment by zeripath** [comment]
Use certificate changes have been removed.

File: services/auth/source/ldap/source_search.go (hunk 1 above)
Comment (zeripath) [reply services/auth/source/ldap/source_search.go]: This functionality has been removed

**Comment by zeripath** [comment]
@lafriks please review again

File: services/auth/source/smtp/auth.go [hunk 2]
```diff
@@ -42,40 +44,63 @@ func (auth *loginAuthenticator) Next(fromServer []byte, more bool) ([]byte, erro
...
+	conn, err := net.Dial("tcp", net.JoinHostPort(source.Host, strconv.Itoa(source.Port)))
 	if err != nil {
 		return err
 	}
-	defer c.Close()
+	defer conn.Close()
 
-	if err = c.Hello("gogs"); err != nil {
-		return err
+	isSecureConn := source.ForceSMTPS || source.Port == 465
```
Comment (lafriks) [inline services/auth/source/smtp/auth.go]: You can use source.UseTLS method here, also no need for variable if it's used just once

**Comment by lafriks** [comment]
Small nit but otherwise looks good

**Review by lafriks — APPROVED** [review]

**Review by 6543 — APPROVED** [review]

**Pull request merged**